internal/upstreamldap/upstreamldap.go

```diff
@@ -0,0 +1,386 @@
+	}
+
+	conn, err := p.dial(ctx)
+	if err != nil {
+		return nil, false, fmt.Errorf(`error dialing host "%s": %w`, p.c.Host, err)
+	}
+	defer conn.Close()
+
+	err = conn.Bind(p.c.BindUsername, p.c.BindPassword)
+	if err != nil {
+		return nil, false, fmt.Errorf(`error binding as "%s" before user search: %w`, p.c.BindUsername, err)
+	}
+
+	mappedUsername, mappedUID, err := p.searchAndBindUser(conn, username, bindFunc)
+	if err != nil {
+		return nil, false, err
+	}
+	if len(mappedUsername) == 0 || len(mappedUID) == 0 {
+		// Couldn't find the username or couldn't bind using the password.
+		return nil, false, nil
+	}
+
+	response := &authenticator.Response{
+		User: &user.DefaultInfo{
+			Name:   mappedUsername,
+			UID:    mappedUID,
+			Groups: []string{}, // Support for group search coming soon.
+		},
+	}
+	return response, true, nil
+}
+
+func (p *Provider) validateConfig() error {
+	if p.c.UserSearch.UsernameAttribute == distinguishedNameAttributeName && len(p.c.UserSearch.Filter) == 0 {
+		// LDAP search filters do not allow searching by DN, so we would have no reasonable default for Filter.
+		return fmt.Errorf(`must specify UserSearch Filter when UserSearch UsernameAttribute is "dn"`)
+	}
+	return nil
+}
+
+func (p *Provider) searchAndBindUser(conn Conn, username string, bindFunc func(conn Conn, foundUserDN string) error) (string, string, error) {
+	searchResult, err := conn.Search(p.userSearchRequest(username))
+	if err != nil {
+		return "", "", fmt.Errorf(`error searching for user "%s": %w`, username, err)
+	}
+	if len(searchResult.Entries) == 0 {
+		plog.Debug("error finding user: user not found (if this username is valid, please check the user search configuration)",
+			"upstreamName", p.GetName(), "username", username)
+		return "", "", nil
+	}
+	if len(searchResult.Entries) > 1 {
+		return "", "", fmt.Errorf(`searching for user "%s" resulted in %d search results, but expected 1 result`,
+			username, len(searchResult.Entries),
+		)
+	}
+	userEntry := searchResult.Entries[0]
+	if len(userEntry.DN) == 0 {
+		return "", "", fmt.Errorf(`searching for user "%s" resulted in search result without DN`, username)
+	}
+
+	mappedUsername, err := p.getSearchResultAttributeValue(p.c.UserSearch.UsernameAttribute, userEntry, username)
+	if err != nil {
+		return "", "", err
+	}
+
+	mappedUID, err := p.getSearchResultAttributeValue(p.c.UserSearch.UIDAttribute, userEntry, username)
+	if err != nil {
+		return "", "", err
+	}
+
+	// Caution: Note that any other LDAP commands after this bind will be run as this user instead of as the configured BindUsername!
+	err = bindFunc(conn, userEntry.DN)
+	if err != nil {
+		plog.DebugErr("error binding for user (if this is not the expected dn for this username, please check the user search configuration)",
+			err, "upstreamName", p.GetName(), "username", username, "dn", userEntry.DN)
+		ldapErr := &ldap.Error{}
+		if errors.As(err, &ldapErr) && ldapErr.ResultCode == ldap.LDAPResultInvalidCredentials {
```

Comment:
One too many `&` on the `ldapErr` (pretty sure `errors.As` accidentally tolerates this though).

Reply:
Taking the `&` off causes this compile error: `second argument to errors.As must be a non-nil pointer to either a type that implements error, or to any interface type`